Content.Client/SubFloor/TrayScannerSystem.cs

```diff
@@ -0,0 +1,176 @@
+
+        scanner.Toggled = toggle;
+        UpdateTrayScanner(uid, scanner);
+
+        if (toggle) _activeScanners.Add(uid);
+    }
+
+    private HashSet<EntityUid> _activeScanners = new();
+    private RemQueue<EntityUid> _invalidScanners = new();
+
+    public override void Update(float frameTime)
+    {
+        if (!_activeScanners.Any()) return;
+
+        foreach (var scanner in _activeScanners)
+        {
+            if (_invalidScanners.List != null
+                && _invalidScanners.List.Contains(scanner))
+                continue;
+
+            if (!UpdateTrayScanner(scanner))
+                _invalidScanners.Add(scanner);
+        }
+
+        foreach (var invalidScanner in _invalidScanners)
+            _activeScanners.Remove(invalidScanner);
+
+        if (_invalidScanners.List != null) _invalidScanners.List.Clear();
+    }
+
+    /// <summary>
+    ///     Updates a T-Ray scanner. Should be called on immediate
+    ///     state change (turned on/off), or during the update
+    ///     loop.
+    /// </summary>
+    /// <returns>true if the update was successful, false otherwise</returns>
+    private bool UpdateTrayScanner(EntityUid uid, TrayScannerComponent? scanner = null, TransformComponent? transform = null)
+    {
+        // whoops?
+        if (!Resolve(uid, ref scanner, ref transform))
+        {
+            return false;
+        }
+
+        // if the scanner was toggled off recently,
+        // set all the known subfloor to invisible,
+        // and return false so it's removed from
+        // the active scanner list
+        if (!scanner.Toggled)
+        {
+            _subfloorSystem.ToggleSubfloorEntities(scanner.RevealedSubfloors, false, uid, _visualizerKeys);
+            scanner.LastLocation = Vector2.Zero;
+            scanner.RevealedSubfloors.Clear();
+            return false;
+        }
+
+        // get the rounded position so that small movements don't cause this to
+        // update every time
+        Vector2 flooredPos;
+
+        // zero vector implies container
+        if (transform.LocalPosition == Vector2.Zero
+            && transform.Parent != null
+            && _containerSystem.ContainsEntity(transform.Parent.Owner.Uid, uid))
+        {
+            flooredPos = transform.Parent.LocalPosition.Rounded();
+
+            // could recurse through fully but i think that's useless,
+            // just attempt to check through the gp's transform and if
+            // that doesn't work, just don't bother any further
+            if (flooredPos == Vector2.Zero)
+            {
+                if (EntityManager.TryGetComponent(transform.Parent.Owner.Uid, out TransformComponent? gpTransform)
+                    && gpTransform.Parent != null
+                    && _containerSystem.ContainsEntity(gpTransform.Parent.Owner.Uid, transform.Parent.Owner.Uid))
+                {
+                    flooredPos = gpTransform.Parent.LocalPosition.Rounded();
+                }
+            }
```

Comment:
Just a small nitpick for the variable names here. You're getting the parent's transform, not the grand-parent's
```suggestion
            // could recurse through fully but i think that's useless,
            // just attempt to check through the grand-parent's position and if
            // that doesn't work, just don't bother any further
            if (flooredPos == Vector2.Zero)
            {
                if (EntityManager.TryGetComponent(transform.Parent.OwnerUid, out TransformComponent? pTransform)
                    && pTransform.Parent != null
                    && _containerSystem.ContainsEntity(pTransform.Parent.OwnerUid, transform.Parent.OwnerUid))
                {
                    flooredPos = pTransform.Parent.LocalPosition.Rounded();
                }
            }
```

Reply:
huh, i thought i:
- resolved the t-ray scanner's transform
- used the transform's parent entity to get the position
- if the position was zero, then attempt to get the parent of that entity's transform

i named it grandparent based on that - however if it's more preferable to focus on the scope of the if block there (which yeah, then it's the parent of that transform) then i won't mind changing it 🦊

Reply:
Actually I've confused myself as well here. its a bit of a mess. I think the confusion stems from the fact that `transform.Parent` is another transform, not an entity, leading to some circular references. I think this is right, you have:
- The scanner (`uid`/`uid`)
- The scanner's transform (`transform`)
- The scanner's parent's transform (`transform.Parent`) **(not the entity)**
- The scanner's parent entity (`transform.ParentUid` OR  `transform.Parent.OwnerUid` OR `transform.Parent.Owner.Uid`)
- The parent's position (`transform.Parent.LocalPosition()`)

You then use `EntityManager.TryGetComponent(transform.Parent.Owner.Uid, out TransformComponent? gpTransform)`
which is just the same as `transform.Parent` so the tryGet entity is completely unnecessary

`transform.Parent.Parent` is the actual grand-parent transform
To get the actual grand parent entity you can just use `transform.Parent.ParentUid`

I'm guessing using `transform.ParentUid` is generally preferred over `transform.Parent.OwnerUid`